### ISSUE TITLE
Unclear error in output when using lambda-local in unit tests

### ISSUE THREAD

**Issue by pizzurro** [ashiina/lambda-local#99]
I have recently started to use lambda-local for unit tests and am a little confused by what the error is in the following output:

```
Lambda Category handler should return the correct number of categories: 
error: Error
error: ------
Assertion {
  __flags: 
   { ssfi: [Function: proxyGetter],
     lockSsfi: undefined,
     object: 8,
     message: null } }
  √ Lambda Category handler should return the correct number of categories: 8ms
  √ Color lambda should return a list of colors: 4ms

  2 passing (19ms)
```

**Comment by hoegertn** [comment]
can you provide the code that is run?
It is very hard to guess what is happening here.

**Comment by pizzurro** [comment]
Sure. This is the code that creates the message:
```
describe('Lambda Category handler', function() {
  it('should return the correct number of categories', (done) => {
    const lambdaFunc = require("../index");
    lambdalocal.setLogger(winston);
    lambdalocal.execute({
      event: jsonPayload,
      lambdaFunc: lambdaFunc, //We are directly passing the lambda function
      lambdaHandler: "getCategories",
      callbackWaitsForEmptyEventLoop: true,
      timeoutMs: 3000,
      mute: true,
      callback: function (err: any, cb: any) { 
        //We are storing the results and finishing the before() call => one lambda local call for multiple tests
        should.not.exist(err);
        expect(cb).to.be.an('array');
        expect(cb).to.have.length(7);
        let firstNode = cb[0];
        expect(firstNode).to.exist;
        expect(firstNode).to.have.all.keys([
          'id',
          'name'
        ]);

        done();
      }
    });
  });
});
```

**Comment by gpotter2** [comment]
I do find the error log quite easy to understand.

```
Lambda Category handler should return the correct number of categories: 
```

Mocha stuff...

```
error: Error
error: ------
```

Beginning of lambda-local logs:
An error has occurred.

```
Assertion {
  __flags: 
   { ssfi: [Function: proxyGetter],
     lockSsfi: undefined,
     object: 8,
     message: null } }
```
Error log.
End of lamba-local logs.
```
  √ Lambda Category handler should return the correct number of categories: 8ms
  √ Color lambda should return a list of colors: 4ms

  2 passing (19ms)
```

End of mocha logs.

As you can see, your tests are working: the callback is working, the error is returned, which is what your test checks. Lambda-local has only also printed the error.

If you want lambda-local not to print the error log, you might mute the Winston module that you pass to lambda-local.
There are many different ways of doing this (creating an empty Winston instance, adding a log level below error...)

Hope this answers your question :)

**Comment by pizzurro** [comment]
Thanks for your response @gpotter2 but there is no error. Are you can see from the output there are no failing tests. That strange assertion message shows up regardless of assertion.

**Comment by gpotter2** [comment]
Are you sure that you are using the latest version ?
Can you log `err` in the callback ?
(PS: you wrote the callback took `err` Andy `cb` but it's `err` and `msg`)

**Comment by pizzurro** [comment]
I am using version 1.4.2
I figured out the weird assertion message.  -- I inadvertently left in `console.log(cb.length.should.equal(8));`  However, even after removing it, changing the second param from "cb" to "msg", and logging "err." I still have an two lines of 'error:' with no error. 

```
  Lambda Category handler
error: Error
error: ------
Value of err:  undefined
    √ should return the correct number of categories (114ms)

  1 passing (124ms)

```

This is the code I am running:
```
var chai = require("chai");
var lambdalocal = require("lambda-local");
var should = chai.should();
var jsonPayload = {};
var winston = require("winston");
winston.level = 'error';
describe('Lambda Category handler', function () {
    it('should return the correct number of categories', function (done) {
        var lambdaFunc = require("../index");
        lambdalocal.setLogger(winston);
        lambdalocal.execute({
            event: jsonPayload,
            lambdaFunc: lambdaFunc,
            lambdaHandler: "getCategories",
            callbackWaitsForEmptyEventLoop: true,
            timeoutMs: 3000,
            mute: true,
            callback: function (err, msg) {
                should.not.exist(err); 
                console.log("Value of err: ", err);
                done();
            }
        });
    });
});
```
There are no errors being raised (if there were, the test would fail).  So, I am not sure why there are what look like error headers being printed without errors.

**Comment by gpotter2** [comment]
Could you please retry with the github version ?
Otherwise it seems like a bug..

**Comment by pizzurro** [comment]
I have the HEAD of the dev branch installed:
```
$ npm list lambda-local
lambda-sample@1.0.0 C:\Users\cp\repo\node-sample\lambda
`-- lambda-local@1.4.2  (git+https://github.com/ashiina/lambda-local.git#f95dccba4a46dc059aaff11235728b427c9beef8)```
```
The output still contains:
```
error: Error
error: ------
```

I removed Winston as the logger allowing the default logger to work.  This changed the message to:
```
info: END
error: Error
error: ------
info:
```
Turns out the problem is here: https://github.com/ashiina/lambda-local/blob/f95dccba4a46dc059aaff11235728b427c9beef8/lib/context.js#L101

`!==` is too strict here as it will evaluate to true when err is undefined.  So, it probably should be: 

```
typeof err === "undefined" || err !== null  // unambiguous intent
```
 or

```
err != null // performs the type coercion but is evil 
```

**Comment by hoegertn** [comment]
what about: `if (err) {`

**Comment by pizzurro** [comment]
you could do `if(!err)` but if false were returned for some reason, this woild evaluate to true and print the error.

**Comment by pizzurro** [comment]
Let me rephrase, if err = false the error header would be printed.  This is probably ok since implementors that are follow node callback patterns shouldn't be setting err to anything unless there is an error.

**Comment by gpotter2** [comment]
I'm ok with
`err != null`

Every one's ok with that ?

**Comment by pizzurro** [comment]
The good news is that the error message is not longer printed.  Unfortunately, the response (formerly suppressed by the error defect) is now being printed to the console even though mute is set to true.  I would assume that muting should be respected throughout the execution but the following line turns it off: https://github.com/ashiina/lambda-local/blob/b63d52b25bf54d5d27e13e4720bf329a02fa31ac/lib/context.js#L96  

Not being able to turn this off makes test output verbose.  Would be nice to have the ability to set logging level.  At minimum, it would be great if `mute === true` turned all logging off.
